Raise ArgumentError when docstring's are not a string (or nil)

JonRowe opened the pull request:
This was originallly rspec/rspec-core#3081 it got merged prior to the monorepo migration to get it into 4-0-dev but it has some consquences when run across the entire build surrounding metadata, and needs properly allowing classes (even if I have a preference against them, they are too widly used to "break" in 4 plus theres the surrounding debate about `described_class` we've yet to have) in example groups.

3.99 counterpart https://github.com/rspec/rspec/pull/107

Fixes https://github.com/rspec/rspec/issues/40

JonRowe committed:
[core] Merge pull request rspec/rspec-core#3081 from eisukeyeongjo/argument_error_when_doc_string_is_not_nil_or_stirng

Raise ArgumentError when doc_string is neither string nor nil

---
This commit was imported from https://github.com/rspec/rspec-core/commit/d8c5e3ec08302607c91fbc45e552e4254e1fe997.

pirj commented on rspec-core/Changelog.md:
```diff
@@ -35,6 +35,10 @@ Breaking Changes:
 * Change the default order to random. (Santiago Bartesaghi, rspec/rspec-core#2929)
+* Raise an error when an example group description is neither a string, class or module.
+  (Eisuke Yoshida, rspec/rspec-core#3081)
+* Raise an error when an example description is nor a string.
```
```suggestion
* Raise an error when an example description is not a string nor a class.
```

pirj committed:
Rework own specs to adhere to new docstring convention

Fix rspec-core specs to adhere to new docstring convention
Fix rspec-core features to adhere to new docstring convention
Adjust features to use String/Class as example group docstring
Rewrite so it handles Ruby 3.4
  1.8.7 spits an "unquoted { in regexp", 3.4 has changed the inspect format
  to {foo: :bar} from {:foo => :bar}

Fix specs example descriptions

Appease RuboCop

pirj commented:
Green

pirj commented:
Green, squashed.

JonRowe commented on rspec-core/lib/rspec/core/example_group.rb:
```diff
@@ -270,6 +274,10 @@ def self.define_example_group_method(name, metadata={})
             args << combined_metadata
 
+            unless NilClass === description || String === description || Class === description || Module === description
+              raise ArgumentError, "Example groups must be described with a string, got: `#{description.inspect}`"
```
```suggestion
              raise ArgumentError, "Example groups must be described with a string, class or module got: `#{description.inspect}`"
```

JonRowe commented on rspec-core/spec/rspec/core/example_group_spec.rb:
```diff
@@ -55,16 +55,6 @@ def metadata_hash(*args)
-    it 'does not treat the first argument as a metadata key even if it is a symbol' do
-      group = RSpec.describe(:symbol)
-      expect(group.metadata).not_to include(:symbol)
-    end
```
These should probably be treated as metadata...

JonRowe commented on rspec-core/spec/rspec/core/example_group_spec.rb:
```diff
@@ -1233,6 +1223,66 @@ def extract_execution_results(group)
+      it "raises ArgumentError when a Hash is used as an example group doc string" do
+        expect { RSpec.describe(foo: :bar) { } }.
+          to raise_error(%r{Example groups must be described with a string, got: `#{Regexp.escape({ foo: :bar }.inspect)}`})
+      end
```
No this is incorrect, this should be interepreted as a `nil` description with metadata, I think that should still be supported for now, just because their are a tonne of aliases and things where that makes sense.

Think `describe type: :model { ... }` for example.

JonRowe commented on rspec-core/spec/rspec/core/example_group_spec.rb:
```diff
@@ -1233,6 +1223,66 @@ def extract_execution_results(group)
+
+      it "raises ArgumentError when a Hash is used as an example doc string" do
+        expect { group.it(foo: :bar) { } }.
+          to raise_error(ArgumentError, %r{Examples must be described with a string, got: `#{Regexp.escape({ foo: :bar }.inspect)}`})
```
Similarly, this is a `nil` description with metadata, this will be more common...
```
specify(tag: :validations) { is_expected.to be_valid }
```

JonRowe commented on rspec-core/spec/rspec/core/example_group_spec.rb:
```diff
@@ -1233,6 +1223,66 @@ def extract_execution_results(group)
+      it "raises ArgumentError when a Symbol is used as an example group doc string" do
+        expect { RSpec.describe :foo }.
+          to raise_error(ArgumentError, /Example groups must be described with a string, got: `:foo`/)
+      end
```
Technically `:foo` here should be considered `foo: true` and be valid as metadata.

JonRowe reviewed (commented):
Thanks for your work here @pirj but this still doesn't tackle the bare "metadata" cases, see comments.

pirj commented:
I see what you mean, @JonRowe

One thing is that any symbol can be treated as metadata, while in most cases it’s not.
Sometimes, it will be harmless, but when this matches some metadata trigger, it can cause surprising behaviour.

In the same way, since most Ruby hashes have symbol keys anyway, and since the same is our deciding factor if it is metadata or not, we can mistakenly take some hash from the first argument and treat it as metadata, while it is not.

I’m aways from my computer and can’t take a closer look now.
However, from memory, even thiugh I would like to formalise this so the first argument should always be a string (class, module). But if we’d introduce some exceptions to this rule, and add detection by type, I would rather not do this at all and kept things as they are.
